cranelift-codegen/src/ir/function.rs

```diff
@@ -118,6 +127,20 @@ impl Function {
         Self::with_name_signature(ExternalName::default(), Signature::new(CallConv::Fast))
     }
 
+    /// Creates a constant in the function.
+    pub fn create_constant(&mut self, constant: impl Into<Constant>) -> bool {
+        let c = constant.into();
+        let d = if self.constants.contains(&c) {
+            c // offset has already been set at previous insertion point
+        } else {
+            c.offset(byte_len(&self.constants) as u32) // set offset at the current number of bytes in the set; relies on the set not being re-ordered
+                                                       // TODO if we want to set the offset here we probably don't want a b-tree which would invalidate earlier offsets
+                                                       // if something was inserted later (in time) that is put earlier (in sort order); we probably want a set
+                                                       // that simply maintains insert order
+        };
+        self.constants.insert(d)
+    }
```

Comment:
If we cannot set the offsets in `relaxation.rs` then it could be done here. The issue is that no one is calling `create_constant` and I am not sure who would; cranelift already knows how to convert constants to `Uimm128` so this code would never be called and the offset never set. Perhaps if we had a separate pass that picked out all of the specific constant types (e.g. `Uimm128`), converted them to the generic `Constant`, and added them to the set of constants, keeping track of their offsets as we go?

Reply:
or add a separate table for constants, just like jumptables.